pex/scie/__init__.py

```diff
@@ -0,0 +1,175 @@
+# Copyright 2024 Pex project contributors.
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import absolute_import
+
+from argparse import Namespace, _ActionsContainer
+
+from pex.fetcher import URLFetcher
+from pex.orderedset import OrderedSet
+from pex.pep_440 import Version
+from pex.scie import science
+from pex.scie.model import (
+    ScieConfiguration,
+    ScieInfo,
+    ScieOptions,
+    SciePlatform,
+    ScieStyle,
+    ScieTarget,
+)
+from pex.typing import TYPE_CHECKING, cast
+from pex.variables import ENV, Variables
+
+if TYPE_CHECKING:
+    from typing import Iterator, Optional, Tuple, Union
+
+
+__all__ = (
+    "ScieConfiguration",
+    "ScieInfo",
+    "SciePlatform",
+    "ScieStyle",
+    "ScieTarget",
+    "build",
+)
+
+
+def register_options(parser):
+    # type: (_ActionsContainer) -> None
+
+    parser.add_argument(
+        "--scie",
+        dest="scie_style",
+        default=None,
+        type=ScieStyle.for_value,
+        choices=ScieStyle.values(),
+        help=(
+            "Create one or more native executable scies from your PEX that include a portable "
+            "CPython interpreter along with your PEX making for a truly hermetic PEX that can run "
+            "on machines with no Python installed at all. If your PEX has multiple targets, "
+            "whether `--platform`s, `--complete-platform`s or local interpreters in any "
+            "combination, then one PEX scie will be made for each platform, selecting the latest "
+            "compatible portable CPython interpreter. Note that only CPython>=3.8 is supported. If "
+            "you'd like to explicitly control the target platforms or the exact portable CPython "
+            "selected, see `--scie-platform`, `--scie-pbs-release` and `--scie-python-version`. "
+            "Specifying `--scie {lazy}` will fetch the portable CPython interpreter just in time "
+            "on first boot of the PEX scie on a given machine if needed. The URL(s) to fetch the "
+            "portable CPython interpreter from can be customized by exporting the "
+            "PEX_BOOTSTRAP_URLS environment variable pointing to a json file with the format: "
+            '`{{"ptex": {{<file name 1>: <url>, ...}}}}` where the file names should match those '
+            "found via `SCIE=inspect <the PEX scie> | jq .ptex` with appropriate replacement URLs. "
+            "Specifying `--scie {eager}` will embed the portable CPython interpreter in your PEX "
+            "scie making for a larger file, but requiring no internet access to boot. If you have "
+            "customization needs not addressed by the Pex `--scie*` options, consider using "
+            "`science` to build your scies (which is what Pex uses behind the scenes); see: "
+            "https://science.scie.app.".format(lazy=ScieStyle.LAZY, eager=ScieStyle.EAGER)
```

Comment:
@sureshjoshi I see that your Pants plugin accepts an optional custom lift manifest, parses it if present, then injects bits into it. I think to support that sort of thing in a principled way, I'd have to parse the user supplied manifest and confirm they do not set the following keys:
+ ptex
+ scie_jump
+ files: with matching names
+ interpreters: with matching ids
+ commands: with a default command (I use this to launch the PEX)
+ bindings: with a matching name (needed for the default command to work)

Additionally, I'd have to advertise that I bind ptex to "ptex" for lazy scies, and bind `configure:PYTHON` and `configure:PEX`.

Without all this I don't see how the user supplied manifest can work with Pex needs fruitfully. Can you think of any other corners? Perhaps I'm overthinking. Do you need this functionality?

Reply:
I guess for `ptex` and `scie_jump` I could allow user-specified versions (but no more) IFF those versions were compatible with a lower bound.

Reply:
I was hacking around tonight, trying to envision how I'd re-build something like `pantsible` (for example). 

One idea was to manipulate the embedded manifest after `pex` generates it (add the custom bindings and whatnot by piping the file to another tool), but then I realized I don't think I'd want to be able to dynamically modify the manifest of what should be a "sealed" binary, as that would be crazy for supply chain purposes - and I don't want to be able to dynamically alter the commands the executable could call.

In the case of the plugin (which, I wouldn't really use as a reference for anything - as I made it a few years ago to solve an immediate deployment problem on a client project), I think we try to use the optional lift.toml where possible and inject the target names under certain conditions. 

For this PR, I don't see any problems with deferring all of those concerns, but I'm of two minds.
- `pex` being able to accept a custom manifest template that has to be perfectly structured, with/without certain keys feels a bit hacky
- Using a separate tool (`science`), which overlaps with a lot of what `pex` would provide, feels off too

Would it make sense/be possible for `science` to defer to `pex` in some way, for the embedded Python interpreter? I'm trying to envision some sort of cleaner composition between two tools which have similar base functionality - but `science` allows some added knobs.

```toml
[lift]
name = "pantsible"
description = "Ansible with an embedded Python interpreter."
platforms = ... inferred from pex ...

[[lift.interpreters]] -> ... inferred from pex ...

[[lift.files]]
name = "pex"

[[lift.commands]]
name = "ansible"
exe = "{scie.bindings.venv}/venv/bin/ansible"
args = []

...
```

Although, one immediate problem I see here... I think I'm conflating a `pex` file with the `pex` CLI.

Reply:
Reading through the PR, another thought that popped into my head is allowing for the `pex` CLI's generated TOML to act as an overlay or merge-manifest with a local one. 

Whether that functionality is in `science` or `pex` CLI - overlaying/overwriting the user created manifest seems reasonable.

Reply:
> I was hacking around tonight, trying to envision how I'd re-build something like pantsible (for example).

Well, `pantsible` uses a feature specific to scies over and above a PEX, namely the busybox support. It makes sense to me to just directly support this with `--scie-busybox [list of entry points]`. If you specify that then Pex emits a manifest with no default command and just named commands for each listed entry point.

> Would it make sense/be possible for science to defer to pex in some way, for the embedded Python interpreter? I'm trying to envision some sort of cleaner composition between two tools which have similar base functionality - but science allows some added knobs.

Well, science is general purpose - Any language; so it doesn't really make sense for it to know about Python let alone Pex. It does have a Provider interface to supply interpreters and that has exactly 1 implementation currently, that provides PBS interpreters. A PEX provider might make sense.

That said, Pex creates PEXes - single file executables. These do not have:
1. BusyBox support: You need [conscript](https://pypi.org/project/conscript/), for example, for that.
2. Bindings support: I.E.: Pex offers you no way to do pre-launch setup. You just have to write Python code to do 1 time setup in your main if you want that or provide alternate entry points fired off with `{PEX_MODULE=foo,PEX_SCRIPT=bar} ./my.pex`
As such, I think it makes sense for Pex to offer the ability to take your PEX file and turn it into a scie that behaves _exactly_ the same, with nothing extra except maybe running faster.  Everything you'd do in a custom manifest, afaict, would add things the PEX _cannot_ already do. At that point, having to move up a layer and use science yourself with a custom lift manifest to buold your app not using Pex directly makes sense. I.E.: what scie-pants has to do. The Pants app is more complex than just what the Pants PEX does / has tight perf overhead concerns; so it makes sense to move up to the higher layer.

> Reading through the PR, another thought that popped into my head is allowing for the pex CLI's generated TOML to act as an overlay or merge-manifest with a local one.

That's exactly what I meant by all this: https://github.com/pex-tool/pex/pull/2466#discussion_r1678670539 It seems to me you can't just overlay, you must confirm the key mechanisms Pex uses in its lift are not destroyed by the merge before merging.

Reply:
> It seems to me you can't just overlay, you must confirm the key mechanisms Pex uses in its lift are not destroyed by the merge before merging.

Ah okay, I thought that overlaying pex's generated output on top of a user-supplied one (wiping out whatever is already there) would work - but I guess there are a lot of edge cases that might crop up along the way.

> At that point, having to move up a layer and use science yourself with a custom lift manifest to build your app not using Pex directly makes sense. I.E.: what scie-pants has to do. The Pants app is more complex than just what the Pants PEX does / has tight perf overhead concerns; so it makes sense to move up to the higher layer.

Agreed. This is what my brain was converging to by the time I stopped reviewing last night. 

The reason why I mentioned what is ultimately a "PEX PBS provider" is to hopefully ease the problem of pex `platforms`/constraints conflicting with the `platforms` and `interpreters` inside of a `science` lift.toml. Or at least, some way to deduce at runtime the valid Python interpreters for the given pex file(s).

Reply:
> Ah okay, I thought that overlaying pex's generated output on top of a user-supplied one (wiping out whatever is already there) would work - but I guess there are a lot of edge cases that might crop up along the way.

Yeah - just blindly over-writing what a user deemed important seems like a recipe for disaster.

> The reason why I mentioned what is ultimately a "PEX PBS provider" is to hopefully ease the problem of pex platforms/constraints conflicting with the platforms and interpreters inside of a science lift.toml. Or at least, some way to deduce at runtime the valid Python interpreters for the given pex file(s).

The current PR does derive the platforms and interpreters from the Pex input by default, you commonly only need to override with --scie-platform, --scie-pbs-release, --scie-python-version if you want:
1. To create less scies than the PEX's natiral set of platform would indicate (I have a test here I'll be pushing shortly that exercises this platform restricting use).
2. To make a reproducible scie: By default a Python version is picked using `<major>.<minor>` which leaves open the possibility PBS does a new patch release and when you re-build the PEX scie later, you get that new patch release and not the current one. In some use scenarios - not cool.

I'm not sure if I'm missing your point on this one or you just didn't see that ecisting functionality in the code / tests.

Reply:
I'm referring to downstream tools like `science`, not pex, in this case. As in "once you've created a pex, then ..."

Anyways, the things I have in my mind are probably out of scope of this PR, and if they're important enough, or strongly enough use-cased, I can open a new ticket later.

Reply:
Gotcha. So I think the PEX interpreter Provider would just use the `pex3 scie create ...` logic I referenced here: https://github.com/pex-tool/pex/pull/2466#discussion_r1678497485

I.E.: not create the scie, but use the `ScieConfiguration.from_tags` API to implement platform / interpreter selection via the calculated `ScieConfiguration`'s `ScieTarget` targets which include platform, pbs_release and python_version.

That said, the current science Provider interface only allows providing an interpreter and not a set of platforms; so new API work would need to be done in science anyhow it seems to plug all this in.

Reply:
I guess the current API does allow enough for a PEX interpreter Provider to error when asked to produce an interpreter distribution via `Provider.distribution(platform)` for a platform the PEX does not support. That's probably actually enough:

```toml
[lift]
name = "example"
platforms = [
    "linux-aarch64",
    "linux-x86_64",
    "macos-aarch64",
]

[[lift.files]]
name = "pex"

[[lift.interpreters]]
id = "cpython"
provider = "PEX"
pex = "{pex}"
```

Here if I ran `science lift --file pex=my-py37.pex build ...` the PEX interpreter Provider could fail since CPython 3.7 is not supported and if I ran `science lift --file pex=my-py38.pex build ...` it could fail fast if, for example, there were no linux-aarch64 distributions in the latest PBS release.

Reply:
Yep, there we go - that's the kinda thing I see value in.

Reply:
The restricting use case for `--scie-platform` I mentioned now has a test in 61f55a4 as does auto platforms detection.